node/src/database/db_migrations/migrations/migration_3_to_4.rs

```diff
@@ -12,8 +12,53 @@ use tiny_hderive::bip32::ExtendedPrivKey;
 #[allow(non_camel_case_types)]
 pub struct Migrate_3_to_4;
 
+impl Migrate_3_to_4 {
+    fn maybe_exchange_seed_for_private_key(
+        consuming_path_opt: Option<String>,
+        example_encrypted_opt: Option<String>,
+        seed_encrypted_opt: Option<String>,
+        utils: &dyn DBMigDeclarationUtilities,
+    ) -> Option<String> {
+        match (
+            consuming_path_opt,
+            example_encrypted_opt,
+            seed_encrypted_opt,
+        ) {
+            (Some(consuming_path), Some(example_encrypted), Some(seed_encrypted)) => {
+                let password_opt = utils.db_password();
+                if !DbEncryptionLayer::password_matches(&password_opt, &Some(example_encrypted)) {
+                    panic!("Migrating Database from 3 to 4: bad password");
+                }
+                let seed_encoded =
+                    DbEncryptionLayer::decrypt_value(&Some(seed_encrypted), &password_opt, "seed")
+                        .expect("Internal error")
+                        .expect("Internal error");
+                let seed_data = decode_bytes(Some(seed_encoded))
+                    .expect("Internal error")
+                    .expect("Internal error");
+                let extended_private_key =
+                    ExtendedPrivKey::derive(seed_data.as_ref(), consuming_path.as_str())
+                        .expect("Internal error");
+                let private_key_data = PlainData::new(&extended_private_key.secret());
+                Some(
+                    Bip39::encrypt_bytes(
+                        &private_key_data.as_slice(),
+                        password_opt.as_ref().expect("Password somehow disappeared"),
+                    )
+                    .expect("Internal error: encryption failed"),
+                )
+            }
+            (None, None, None) => None,
```

Comment:
I believe it's possible to have the database contain a password (so that `example_encrypted_opt` is populated) but not yet have wallets generated (so that the path and the seed are not populated) and still be in a consistent state.